Example/WalletConnect/WCSessionViewController.swift

```diff
@@ -182,3 +213,71 @@ class ViewController: UIViewController {
         approve(accounts: [address], chainId: chainId)
     }
 }
+
+extension WCSessionViewController {
+    func applicationDidEnterBackground(_ application: UIApplication) {
+        print("<== applicationDidEnterBackground")
+
+        if interactor?.state != .connected {
+            return
+        }
+
+        if notificationGranted {
+            pauseInteractor()
+        } else {
+            startBackgroundTask(application)
+        }
+    }
+
+    func applicationWillEnterForeground(_ application: UIApplication) {
+        print("==> applicationWillEnterForeground")
+        if let id = backgroundTaskId {
+            application.endBackgroundTask(id)
+        }
+        backgroundTimer?.invalidate()
+
+        if recoverSession {
+            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(500)) {
+                self.interactor?.resume()
+            }
+        }
+    }
+
+    func startBackgroundTask(_ application: UIApplication) {
+        backgroundTaskId = application.beginBackgroundTask(withName: "WalletConnect", expirationHandler: {
+            self.backgroundTimer?.invalidate()
+            print("<== background task expired")
+        })
+
+        var alerted = false
+        backgroundTimer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { _ in
+            print("<== background time remainning: ", application.backgroundTimeRemaining)
```

Comment:
does this trigger background timer to fire many times when you come up? we had that issue before

Reply:
we have code in `applicationWillEnterForeground`

```swift
if let id = backgroundTaskId {
    application.endBackgroundTask(id)
}
backgroundTimer?.invalidate()
```

should be enough?